tensorflow-frameworks/tensorflow-data/src/main/java/org/tensorflow/data/DatasetIterator.java

```diff
@@ -0,0 +1,214 @@
+import org.tensorflow.op.Ops;
+import org.tensorflow.tools.Shape;
+
+import java.util.List;
+
+/**
+ * Represents the state of an iteration through
+ * a tf.data Datset.
+ * <p>
+ * Example: Iteration in graph mode.
+ *
+ * <pre>{@code
+ *  // Create input tensors
+ *  Operand<?> XTensor = tf.constant( ... );
+ *  Operand<?> yTensor = tf.constant( ... );
+ *
+ *
+ *  Dataset dataset = Dataset
+ *          .fromTensorSlices(XTensor, yTensor);
+ *          .batch(BATCH_SIZE);
+ *
+ *  DatasetIterator iterator = dataset.makeIterator();
+ *  List<Output<?>> components = iterator.getNext();
+ *  Operand<?> XBatch = components.get(0);
+ *  Operand<?> yBatch = components.get(1);
+ *
+ *  // Build a TensorFlow graph that does something on each element.
+ *  loss = computeModelLoss(X, y);
+ *
+ *  optimizer = ... // create an optimizer
+ *  trainOp = optimizer.minimize(loss);
+ *
+ *  try (Session session = new Session(graph) {
+ *      try {
+ *          session.run(trainOp);
+ *          ...
+ *      } catch (IndexOutOfBoundsException e) {
+ *          System.out.println("finished iterating.");
+ *          break;
+ *      }
+ *  }
+ *
+ * }</pre>
+ * <p>
+ * Example: Iteration in eager mode.
+ *
+ * <pre>{@code
+ *  // Create input tensors
+ *  Operand<?> XTensor = tf.constant( ... );
+ *  Operand<?> yTensor = tf.constant( ... );
+ *
+ *  int BATCH_SIZE = ...
+ *
+ *  Dataset dataset = Dataset
+ *          .fromTensorSlices(XTensor, yTensor)
+ *          .batch(BATCH_SIZE);
+ *  DatasetIterator iterator = dataset.makeIterator();
+ *
+ *  Optimizer optimizer = ... // create an optimizer
+ *
+ *  for (List<Output<?>> components : dataset) {
+ *      Operand<?> XBatch = components.get(0);
+ *      Operand<?> yBatch = components.get(1);
+ *
+ *      loss = computeModelLoss(X, y);
+ *      trainOp = optimizer.minimize(loss);
+ *  }
+ * }</pre>
+ */
+class DatasetIterator {
+    public static final String EMPTY_SHARED_NAME = "";
+
+    private Ops tf;
+
+    private Operand<?> iteratorResource;
+    private Op initializer;
+
+    private List<DataType<?>> outputTypes;
+    private List<Shape> outputShapes;
+
+    /**
+     * @param tf               Ops accessor corresponding to the same `ExecutionEnvironment`
+     *                         as the `iteratorResource`.
+     * @param iteratorResource An Operand representing the iterator
+     *                         (e.g. constructed from `tf.data.iterator` or
+     *                         `tf.data.anonymousIterator`)
+     * @param initializer      An `Op` that should be run to initialize this iterator
+     * @param outputTypes      A list of `DataType` objects corresponding to the
+     *                         types of each component of a dataset element.
+     * @param outputShapes     A list of `Shape` objects corresponding to the
+     *                         shapes of each componenet of a dataset element.
+     */
+    public DatasetIterator(Ops tf, Operand<?> iteratorResource,
```

Comment:
Should these constructors be public? They don't validate that the iteratorResource is actually an iterator.

Reply:
I agree, makes sense to make them private for now, and require the use of `makeInitializer`. Updated those.